<div style="text-align: center; padding: 20px;">
    <img src="austin-distel-744oGeqpxPQ-unsplash.jpeg" alt="​​Subscription Renewal Insights for a SaaS Company" width="450"/>
</div>


<!-- Image source - https://unsplash.com/photos/person-using-macbook-pro-744oGeqpxPQ
-->

    
A SaaS company seeks to uncover what drives its clients to renew subscriptions. They’ve collected data on client details, subscription records, and economic indicators and would like to connect them to better understand its clients’ behavior. 

They’ve tasked you with analyzing these datasets to identify the key factors influencing clients’ decisions to renew their subscriptions. 

Your analysis will provide them with insights into which customers are renewing their products and the reasons behind their renewals. The company can leverage these insights to make informed decisions to increase renewal rates and improve customer loyalty, helping them stay competitive and ensure long-term growth.


## The Data

The company have provided you with three datasets for your analysis. A summary of each data is provided below.

## `client_details.csv`

| Column         | Description|
|----------------|---------------------------------------------------------------|
| `client_id`    | Unique identifier for each client. |
| `company_size` | Size of the company (Small, Medium, Large).|
| `industry`     | Industry to which the client belongs (Fintech, Gaming, Crypto, AI, E-commerce).|
| `location`     | Location of the client (New York, New Jersey, Pennsylvania, Massachusetts, Connecticut).|

## `subscription_records.csv`

| Column             | Description   |
|--------------------|---------------|
| `client_id`        | Unique identifier for each client.|
| `subscription_type`| Type of subscription (Yearly, Monthly).|
| `start_date`       | Start date of the subscription - YYYY-MM-DD.|
| `end_date`         | End date of the subscription - YYYY-MM-DD.|
| `renewed`          | Indicates whether the subscription was renewed (True, False).|

## `economic_indicators.csv`

| Column           | Description                                       |
|------------------|---------------------------------------------------|
| `start_date`     | Start date of the economic indicator (Quarterly) - YYYY-MM-DD.|
| `end_date`       | End date of the economic indicator (Quarterly) - YYYY-MM-DD.|
| `inflation_rate` | Inflation rate in the period.|
| `gdp_growth_rate`| Gross Domestic Product (GDP) growth rate in the period.|


In [131]:
# Re-run this cell
# Import required libraries
import pandas as pd

# Import data
client_details = pd.read_csv('data/client_details.csv')
subscription_records = pd.read_csv('data/subscription_records.csv', parse_dates = ['start_date','end_date'])
economic_indicators = pd.read_csv('data/economic_indicators.csv', parse_dates = ['start_date','end_date'])

In [132]:
print(client_details.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   client_id     100 non-null    int64 
 1   company_size  100 non-null    object
 2   industry      100 non-null    object
 3   location      100 non-null    object
dtypes: int64(1), object(3)
memory usage: 3.2+ KB
None


In [ ]:
# Start coding here, good luck!
# Data exploration, duplicated values
print(client_details.duplicated().sum())

0


In [ ]:
print(client_details.head())

    client_id company_size    industry       location
0  4280387012        Large     Fintech       New York
1  2095513148        Small     Fintech     New Jersey
2  7225516707       Medium     Fintech   Pennsylvania
3  8093537819        Large      Crypto       New York
4  4387541014       Medium  E-commerce  Massachusetts


In [135]:
print(subscription_records.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   client_id          100 non-null    int64         
 1   subscription_type  100 non-null    object        
 2   start_date         100 non-null    datetime64[ns]
 3   end_date           100 non-null    datetime64[ns]
 4   renewed            100 non-null    bool          
dtypes: bool(1), datetime64[ns](2), int64(1), object(1)
memory usage: 3.3+ KB
None


In [136]:
print(subscription_records.duplicated().sum())

0


In [137]:
print(subscription_records.head())

    client_id subscription_type start_date   end_date  renewed
0  1131383004            Yearly 2020-11-11 2021-11-11    False
1  4309371709           Monthly 2021-05-24 2021-06-23     True
2  3183675157            Yearly 2021-12-25 2022-12-25     True
3  5371694837           Monthly 2020-03-14 2020-04-13     True
4  5157113076           Monthly 2019-11-07 2019-12-07    False


In [138]:
print(economic_indicators.isna().sum())

Unnamed: 0         0
start_date         0
end_date           0
inflation_rate     0
gdp_growth_rate    0
dtype: int64


In [139]:
print(economic_indicators.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21 entries, 0 to 20
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   Unnamed: 0       21 non-null     int64         
 1   start_date       21 non-null     datetime64[ns]
 2   end_date         21 non-null     datetime64[ns]
 3   inflation_rate   21 non-null     float64       
 4   gdp_growth_rate  21 non-null     float64       
dtypes: datetime64[ns](2), float64(2), int64(1)
memory usage: 968.0 bytes
None


In [140]:
print(economic_indicators.shape)

(21, 5)


In [ ]:
# Converting object data types to string
client_details_cols= client_details.select_dtypes(include= 'object').columns
for col in client_details_cols:
    client_details[col]= client_details[col].astype('string')

In [142]:
subscription_records['subscription_type']= subscription_records['subscription_type'].astype('string')

In [143]:
print(client_details.dtypes)

client_id        int64
company_size    string
industry        string
location        string
dtype: object


In [144]:
print(subscription_records.dtypes)

client_id                     int64
subscription_type            string
start_date           datetime64[ns]
end_date             datetime64[ns]
renewed                        bool
dtype: object


In [145]:
print(client_details.shape, subscription_records.shape)

(100, 4) (100, 5)


In [146]:
# Merging dataframes
client_subscription_df= client_details.merge(subscription_records, how= 'inner', on= 'client_id')
print(client_subscription_df.head())

    client_id company_size    industry  ... start_date   end_date renewed
0  4280387012        Large     Fintech  ... 2022-11-25 2023-11-25    True
1  2095513148        Small     Fintech  ... 2021-11-03 2021-12-03   False
2  7225516707       Medium     Fintech  ... 2021-01-19 2022-01-19    True
3  8093537819        Large      Crypto  ... 2019-09-14 2019-10-14   False
4  4387541014       Medium  E-commerce  ... 2018-11-08 2018-12-08   False

[5 rows x 8 columns]


In [147]:
print(client_subscription_df.shape)

(100, 8)


In [148]:
# Function to count the number of Fintech and Crypto clients
def counting(industry):
    if client_subscription_df['industry'].isin([industry]).any():
        response = client_subscription_df[client_subscription_df['industry']== industry].__len__()
        return response

In [149]:
print('Fintech clients:', counting('Fintech'), '  Crypto clients:', counting('Crypto'))

Fintech clients: 22   Crypto clients: 25


In [ ]:
# Total of Fintech and Crypto clients
total_fintech_crypto_clients= counting('Fintech') + counting('Crypto')
print(total_fintech_crypto_clients)

47


In [151]:
print(client_subscription_df.columns)

Index(['client_id', 'company_size', 'industry', 'location',
       'subscription_type', 'start_date', 'end_date', 'renewed'],
      dtype='object')


In [ ]:
# Mapping bool to numerical values 
client_subscription_df['renewed']= client_subscription_df['renewed'].map(int)
print(client_subscription_df.head(3))

    client_id company_size industry  ... start_date   end_date renewed
0  4280387012        Large  Fintech  ... 2022-11-25 2023-11-25       1
1  2095513148        Small  Fintech  ... 2021-11-03 2021-12-03       0
2  7225516707       Medium  Fintech  ... 2021-01-19 2022-01-19       1

[3 rows x 8 columns]


In [ ]:
# Mean of renewed subscriptions by industry
industries= client_subscription_df.groupby('industry')['renewed'].mean()
print(industries)

industry
AI            0.636364
Crypto        0.440000
E-commerce    0.450000
Fintech       0.545455
Gaming        0.727273
Name: renewed, dtype: float64


In [ ]:
# Sorted
industries_sorted= industries.sort_values(ascending= False)
print(industries_sorted)

industry
Gaming        0.727273
AI            0.636364
Fintech       0.545455
E-commerce    0.450000
Crypto        0.440000
Name: renewed, dtype: float64


In [155]:
# Which industry has the highest renewal rate?
top_industry= industries_sorted.index[0]
print(top_industry)

Gaming


In [156]:
print(economic_indicators['start_date'].unique())

['2018-01-01T00:00:00.000000000' '2018-04-01T00:00:00.000000000'
 '2018-07-01T00:00:00.000000000' '2018-10-01T00:00:00.000000000'
 '2019-01-01T00:00:00.000000000' '2019-04-01T00:00:00.000000000'
 '2019-07-01T00:00:00.000000000' '2019-10-01T00:00:00.000000000'
 '2020-01-01T00:00:00.000000000' '2020-04-01T00:00:00.000000000'
 '2020-07-01T00:00:00.000000000' '2020-10-01T00:00:00.000000000'
 '2021-01-01T00:00:00.000000000' '2021-04-01T00:00:00.000000000'
 '2021-07-01T00:00:00.000000000' '2021-10-01T00:00:00.000000000'
 '2022-01-01T00:00:00.000000000' '2022-04-01T00:00:00.000000000'
 '2022-07-01T00:00:00.000000000' '2022-10-01T00:00:00.000000000'
 '2023-01-01T00:00:00.000000000']


In [ ]:
print('Min Start date:', subscription_records['start_date'].min(), '   Max Start date:', subscription_records['start_date'].max())

Min Start date: 2018-02-10 00:00:00    Max Start date: 2022-12-20 00:00:00


In [158]:
subscription_records= subscription_records.sort_values('end_date')
economic_indicators= economic_indicators.sort_values('start_date')

In [159]:
# Merging dataframes subscription_records and economic_indicators
economics_df= pd.merge_asof(subscription_records, economic_indicators, 
                            left_on= 'end_date', right_on= 'start_date', direction= 'backward')
print(economics_df)

     client_id subscription_type  ... inflation_rate gdp_growth_rate
0   4519356806           Monthly  ...           1.17            2.15
1   3683504527           Monthly  ...           1.17            2.15
2   7462725203           Monthly  ...           1.17            2.15
3   3878940490           Monthly  ...           1.17            2.15
4   6751372012           Monthly  ...           1.17            2.15
..         ...               ...  ...            ...             ...
95  7925430606            Yearly  ...           5.49            3.44
96  8884551090            Yearly  ...           5.49            3.44
97  7744303708            Yearly  ...           5.49            3.44
98  4280387012            Yearly  ...           5.49            3.44
99  3411013676            Yearly  ...           5.49            3.44

[100 rows x 10 columns]


In [ ]:
# Not null values
print(economics_df['gdp_growth_rate'].notna().sum())

100


In [ ]:
# Renewed subscriptions filtered
economics_1= economics_df[economics_df['renewed'] == True]
print(economics_1.shape)


(55, 10)


In [ ]:
# Average inflation for renewals
average_inflation_for_renewals= economics_1['inflation_rate'].mean()
print(average_inflation_for_renewals.round(2))

4.42


<b>Actividades:</b>
- Definir una función para contabilizar el número de clientes Fintech y de criptomoneda.
- Obtener el promedio de suscripciones por industria.
- Responder qué industria tiene la tasa de renovación de suscripciones más alta.
- Filtrar un dataframe de las suscripciones que han sido renovadas y obtener el valor de inflación promedio.

<b>Conclusiones:</b>
- Usuarios Fintech y de criptomoneda: <b>47</b>
- Industria con la tasa de renovación más alta: <b>Gaming.</b>
- Valor de inflación promedio en suscripciones con renovación: <b>4.42</b>
